Allow customising key symbols

josmith42 opened the pull request:
* Allows users to customize the symbols used for showing key codes
* Customized symbols are stored in `key_config.ron`
* Example symbols config file added in `assets/alternate_key_symbols.ron`
* Addresses issue [465](https://github.com/extrawurst/gitui/issues/465)

extrawurst commented:
@josmith42 sorry for taking so long to check this out. one thing that bothers me here is that this increases the key.ron file further. how about making the symbols a separate ron file?

josmith42 commented:
@extrawurst Yes, that makes sense. I had actually considered making the symbols a separate file when I did this originally. I don't remember why I decided not to, but I agree that's probably the best way to do it. I'll work on making the change and add another commit to this PR.

extrawurst commented:
I will mark this as draft in the meantime, feel free to switch it back once you think its ready for a round 2

josmith42 commented:
@extrawurst I have put the symbols settings into a separate ron file. I did this by making a separate `SymbolConfig` struct and adding it as a non-serializable field in `KeyConfig`. This seemed like the best way to implement it that wasn't too invasive. Please let me know your feedback. Thank you!

extrawurst commented:
@josmith42 please run `make check` locally before pushing to reduce the turn around times due to broken CI

josmith42 commented:
@extrawurst My apologies. Should be fine now.

extrawurst commented on src/keys.rs:
```diff
@@ -19,6 +19,110 @@ use crate::args::get_app_config_path;
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct SymbolConfig {
+    pub enter_symbol: String,
```
I think we can ditch the `_symbol` postfix considering the struct name

josmith42 replied:
Done.

extrawurst commented on src/keys.rs:
```diff
@@ -81,6 +185,9 @@ pub struct KeyConfig {
     pub abort_merge: KeyEvent,
+
+    #[serde(skip)]
+    symbols: SymbolConfig,
```
what's the reasoning to put this in here and skip in serde even though we use serde to serialise it?

josmith42 replied:
So I did the skip here so that the symbols don't show up in `key_config.ron`, since that file is already getting big. I'm serializing the symbols elsewhere in a separate file.

I'm not terribly fond of this implementation, though. `symbols` has to have default values when the key config is read from `key_config.ron` before it's overwritten with the data in `symbol_config.ron`. The other option I thought of was to make a struct that contains a `KeyConfig` object and a `SymbolConfig` object. That could then be passed around where the current `KeyConfig` struct is. I started going down this route, but this would have ended up touching nearly every file in the code base, because `KeyConfig::get_hint()` would need the data from both objects.

extrawurst commented on assets/alternate_key_symbols.ron:
```diff
@@ -0,0 +1,20 @@
+    up_symbol: "▲",
+    down_symbol: "▼",
+    backspace_symbol: "BACK",
+    home_symbol: "HOME",
+    end_symbol: "END",
+    page_up_symbol: "PgUp",
+    page_down_symbol: "PgDn",
```
Whats the reasoning for ALLCAPS vs. CamelCase?

josmith42 replied:
None, other than I had in my head that it was "standard". Looking at my personal keyboard, though, it's totally not. I'll make it CamelCase.

josmith42 commented:
@extrawurst I addressed the issues you raised in the most recent commit. Please let me know of anything else, and if my explanation of why I implemented it the way I did makes sense.

extrawurst commented:
superseded by https://github.com/extrawurst/gitui/pull/997
